crates/shared/src/order_validation.rs

```diff
@@ -71,11 +71,18 @@ pub trait OrderValidating: Send + Sync {
     ///
     /// Furthermore, full order validation also calls partial_validate to ensure
     /// that other aspects of the order are not malformed.
+    ///
+    /// `full_app_data_override` can be used when the order specifies only a
+    /// contract app data hash without the full app data. In this case
+    /// `full_app_data_override` is used as the full app data and the contract
+    /// app data hash is not validated against it (the hash doesn't have to
+    /// match). The full app data is still otherwise validated.
     async fn validate_and_construct_order(
         &self,
         order: OrderCreation,
         domain_separator: &DomainSeparator,
         settlement_contract: H160,
+        full_app_data_override: Option<String>,
```

Comment:
Instead of an extra field, would it be slightly cleaner to have:

```
        payload.app_data = match payload.app_data {
            OrderCreationAppData::Hash { hash } => if let Some(fetched) = self.find_full_app_data(&hash).await {
                OrderCreationAppData::Full { .. }
            } else {
                OrderCreationAppData::Hash
            },
            other => other,
        };
```

Reply:
The problem with turning Hash into Full is that for full we calculate the hash in the verification logic. But the hash won't match for orders that weren't created with the new app data scheme.

Reply:
I do intuitively think this part can be done cleaner but I'm not sure yet what the best way is.

Reply:
Fine by me.